Source/#NSDockTile.m#

```diff
@@ -0,0 +1,216 @@
+/* Implementation of class NSDockTile
+   Copyright (C) 2019-2023 Free Software Foundation, Inc.
+
+   By: Gregory John Casamento <greg.casamento@gmail.com>
+       Riccardo Mottola <rm@gnu.org>
+
+   Date: Sat Nov 16 21:11:06 EST 2019
+
+   This file is part of the GNUstep Library.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "AppKit/NSApplication.h"
+#import "AppKit/NSDockTile.h"
+#import "AppKit/NSView.h"
+#import "AppKit/NSImage.h"
+#import "AppKit/NSImageRep.h"
+#import "AppKit/NSCustomImageRep.h"
+#import "AppKit/NSFont.h"
+#import "AppKit/NSStringDrawing.h"
+#import "AppKit/NSAttributedString.h"
+#import "AppKit/NSBezierPath.h"
+
+#import "GNUstepGUI/GSDisplayServer.h"
+
+@implementation NSDockTile
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      GSDisplayServer *server = GSCurrentServer();
+      NSSize size = [server iconSize]; 
+      NSRect rect = NSMakeRect(0,0,size.width,size.height);
+      
+      _size = size;
+      _contentView = [[NSView alloc] initWithFrame: rect];
+      _badgeLabel = nil;
+      _owner = nil;
+      _showsApplicationBadge = YES;
+      _appIconImage = [NSImage imageNamed: @"NSApplicationIcon"];
+    }
+  return self;
+}
+
+- (oneway void) release
+{
+  RELEASE(_contentView);
+  RELEASE(_badgeLabel);
+  RELEASE(_appIconImage);
+  [super release];
+}
+
+- (NSView *) contentView
+{
+  return _contentView;
+}
+
+- (void) setContentView: (NSView *)contentView
+{
+  ASSIGN(_contentView, contentView);
+}
+
+- (NSSize) size
+{
+  return _size;
+}
+
+- (id) owner
+{
+  return _owner;
+}
+
+- (void) setOwner: (id)owner
+{
+  _owner = owner; // weak...
+}
+
+- (BOOL) showsApplicationBadge
+{
+  return _showsApplicationBadge;
+}
+
+- (void) setShowsApplicationBadge: (BOOL)flag
+{
+  _showsApplicationBadge = flag;
```

Comment:
This method and setBadgeLabel: should both simply set the properties/ivar.   The display method should be called in each of these methods so that it can re-render and display the icon appropriately.

Reply:
You commented on an outdated version of the code. This does not help as it is hard to say whether this comments are still relevant or not.

Reply:
THe code uses a NSCustomImageRep, thus it does not use the draw method directly, since that one has its own delegate. It is a side-effect of not being able to just "draw" a standard bitmap image rep.

Furthermore setShowsApplicationBadge is strange... apple's Doc says it exists since 10.5, as the DockTile and badge itself. However it says this strange thing:

> In macOS 10.5 and later, application tiles do not support the application badge.

What does this mean? Before 10.5 the badge didn't even exist... maybe application tile != dock tile and this tries to specify this?


and trying (and googling confirms) on my mac, I get this:
`setShowsApplicationBadge: is not yet implemented for the NSApp dockTile`
I don't know if it has been implemented very recently, but apparently up to Big Sur this is the message...